tfhe/src/core_crypto/entities/compressed_modulus_switched_multi_bit_lwe_ciphertext.rs

```diff
@@ -0,0 +1,339 @@
+///
+/// // Decrypt the PBS multiplication result
+/// let decompressed_plaintext: Plaintext<u64> =
+///     decrypt_lwe_ciphertext(&big_lwe_sk, &decompressed_ct);
+///
+/// // Create a SignedDecomposer to perform the rounding of the decrypted plaintext
+/// // We pass a DecompositionBaseLog of 5 and a DecompositionLevelCount of 1 indicating we want to
+/// // round the 5 MSB, 1 bit of padding plus our 4 bits of message
+/// let signed_decomposer =
+///     SignedDecomposer::new(DecompositionBaseLog(5), DecompositionLevelCount(1));
+///
+/// // Round and remove our encoding
+/// let decompression_result: u64 =
+///     signed_decomposer.closest_representable(decompressed_plaintext.0) / delta;
+///
+/// assert_eq!(decompression_result, input_message);
+/// ```
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub struct CompressedModulusSwitchedMultiBitLweCiphertext<
+    Scalar: UnsignedInteger + CastInto<usize> + CastFrom<usize>,
+> {
+    body: usize,
+    packed_mask: PackedIntegers<usize>,
+    lwe_dimension: LweDimension,
+    uncompressed_ciphertext_modulus: CiphertextModulus<Scalar>,
+    grouping_factor: LweBskGroupingFactor,
+}
+
+impl<Scalar: UnsignedInteger + CastInto<usize> + CastFrom<usize>>
+    CompressedModulusSwitchedMultiBitLweCiphertext<Scalar>
+{
+    /// Compresses a ciphertext by reducing its modulus
+    /// This operation adds a lot of noise
+    pub fn compress<Cont: Container<Element = Scalar>>(
+        ct: &LweCiphertext<Cont>,
+        log_modulus: CiphertextModulusLog,
+        grouping_factor: LweBskGroupingFactor,
+    ) -> Self {
+        let uncompressed_ciphertext_modulus = ct.ciphertext_modulus();
+
+        assert!(
+            ct.ciphertext_modulus().is_power_of_two(),
+            "Modulus switch compression doe not support non power of 2 input moduli",
+        );
+
+        let uncompressed_ciphertext_modulus_log =
+            if uncompressed_ciphertext_modulus.is_native_modulus() {
+                Scalar::BITS
+            } else {
+                uncompressed_ciphertext_modulus.get_custom_modulus().ilog2() as usize
+            };
+
+        assert!(
+            log_modulus.0 <= uncompressed_ciphertext_modulus_log,
+            "The log_modulus (={}) for modulus switch compression must be smaller than the uncompressed ciphertext_modulus_log (={})",
+            log_modulus.0,
+            uncompressed_ciphertext_modulus_log,
+        );
+
+        let (input_lwe_mask, input_lwe_body) = ct.get_mask_and_body();
+
+        let body = modulus_switch(*input_lwe_body.data, log_modulus).cast_into();
+
+        let switched_modulus_input_mask_per_group: Vec<_> = input_lwe_mask
+            .as_ref()
+            .chunks_exact(grouping_factor.0)
+            .flat_map(|lwe_mask_elements| {
+                modulus_switch_multi_bit(log_modulus, grouping_factor, lwe_mask_elements)
+            })
+            .collect();
```

Comment:
this does not compress for large grouping factor right ? where are the correction coefficients for the blind rotation we discussed ?

Reply:
I haven't implemented that yet 
I guess we can do it in another PR

Reply:
I would like to have it in this oneI don't think it will be big, it's about comparing the modulus switched values and storing an additional packed integer vector for the corrections

Reply:
the only potential difficulty is how many bits to store, but that can be determined at compression time worst case instead of a generic formula

Reply:
though for conformance likely better to have a higher bound already computed